### PR TITLE
Add --format option to docs command

### LINKED ISSUE #2772: Option for the doc command to generate JSON.
Refs rrrene/inch#38.
/cc @rrrene

### PR DESCRIPTION
**Pull request by mniak**
Fixes #2772 
https://github.com/crystal-lang/crystal/issues/2772#issuecomment-341139030

### CODE REVIEW

File: src/compiler/crystal/tools/doc/generator.cr [hunk 1]
```diff
@@ -28,7 +28,7 @@ class Crystal::Doc::Generator
     },
   }
 
-  def initialize(@program : Program, @included_dirs : Array(String), @output_dir : String, @canonical_base_url : String?)
+  def initialize(@program : Program, @included_dirs : Array(String), @output_dir : String, @output_format : String, sss @canonical_base_url : String?)
```
Comment (Sija) [inline src/compiler/crystal/tools/doc/generator.cr]: `sss`? ;)

Comment (mniak) [reply src/compiler/crystal/tools/doc/generator.cr]: Sorry, typo

File: src/compiler/crystal/command/docs.cr [hunk 2]
```diff
@@ -4,7 +4,10 @@
 # is in `crystal/tools/doc/`
 
 class Crystal::Command
+  private VALID_OUTPUT_FORMATS = ["html", "json"]
```
Comment (Sija) [inline src/compiler/crystal/command/docs.cr]: Pedantic: `%w(html json)`

Comment (straight-shoota) [reply src/compiler/crystal/command/docs.cr]: It's not really an improvement, just a different way to express this.

File: src/compiler/crystal/tools/doc/generator.cr [hunk 3]
```diff
@@ -74,6 +72,25 @@ class Crystal::Doc::Generator
...
+  end
+
+  def generate_docs_html(program_type, types)
+    copy_files
+    generate_types_docs types, @output_dir, types
+    generate_readme program_type, types
+  end
+
+  def generate_readme(program_type, types)
+    raw_body, body = read_readme
```
Comment (straight-shoota) [inline src/compiler/crystal/tools/doc/generator.cr]: The part for generating `body` is only needed in this method, so I suppose it should be moved here from `read_readme`.

**Comment by mniak** [comment]
Should I skip the generation of `index.json` when `--format=html` (or missing)?

**Comment by straight-shoota** [comment]
That would require to change the JavaScript for loading the docs. Currently, it loads `index.json` when served through HTTP and `search-index.js` when opened as a local file (using JSONP). But it's totally fine to use `search-index.js` in both cases.

https://github.com/crystal-lang/crystal/blob/7b8326facbf6ff07b88c8295d749c16776c46545/src/compiler/crystal/tools/doc/html/js/_search.js#L528

You can tackle it right now or leave that for a new PR.

**Comment by mniak** [comment]
I'd prefer to do handle it in a different PR

**Comment by Sija** [comment]
@mniak It would be good to add "to docs command" in the title since it's not clear now what's this option is about.

Also please change "Solves the issue" to "Fixes" or "Closes" since these two are the keywords used by GH to auto-close issues.

**Comment by mniak** [comment]
@Sija thanks for these tips

**Comment by mniak** [comment]
@bcardiff I tried to follow your tip avoiding duplicate the logic to grab the content of the readme file.

**Comment by RX14** [comment]
@bcardiff this is a pretty simple PR to review :)

**Comment by RX14** [comment]
Merging this broke master :(

**Comment by bcardiff** [comment]
A conflict with the specs added during #7063

**Pull request merged**